lib/wasi/src/state/pipe.rs

```diff
@@ -93,26 +234,108 @@ impl WasiPipe {
     }
 }
 
+impl Write for WasiPipe {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let buf_len = buf.len();
+        let tx = self.tx.lock().unwrap();
+        tx.send(buf.to_vec())
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("{e}")))?;
+        Ok(buf_len)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Seek for WasiPipe {
+    fn seek(&mut self, _: SeekFrom) -> io::Result<u64> {
+        Ok(0)
+    }
+}
+
 impl Read for WasiPipe {
     fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
         loop {
             if let Some(inner_buf) = self.read_buffer.as_mut() {
                 let buf_len = inner_buf.len();
                 if buf_len > 0 {
-                    let mut reader = inner_buf.as_ref();
-                    let read = reader.read(buf).map(|_| buf_len as usize)?;
-                    inner_buf.advance(read);
-                    return Ok(read);
+                    if inner_buf.len() > buf.len() {
+                        let mut reader = inner_buf.as_ref();
+                        let read = reader.read_exact(buf).map(|_| buf.len())?;
+                        inner_buf.advance(read);
+                        return Ok(read);
+                    } else {
+                        let mut reader = inner_buf.as_ref();
+                        let read = reader.read(buf).map(|_| buf_len as usize)?;
+                        inner_buf.advance(read);
+                        return Ok(read);
+                    }
+                } else {
+                    return Ok(0);
```

Comment:
Not sure this is the right behavior - if the local pipe is empty it should probably fall through and try and get more data rather than return zero (which indicates end-of-file) as it might not be the case that there is no more data.

Reply:
Also I suspect that if the buffer reaches zero it never read anymore data as it will never get past this point

Reply:
Resolved in https://github.com/wasmerio/wasmer/pull/3145/commits/64399973e33ea3edf8735965c9aa8820a8520dbe

Reply:
No, this has to stay as is. If you fall through, it will go in an indefinite loop with `try_recv()` (in cases where we don't want to block). Maybe I need to add a `set_nonblocking(true / false)` method on the pipe, so that users can configure the behavior.

Reply:
The only question is whether the default option should be blocking or non-blocking. I'm going with blocking for now, but for example for tests, you'd need it to be non-blocking (because CI tests can't wait for keyboard input).

Reply:
Resolved in https://github.com/wasmerio/wasmer/pull/3145/commits/f21689d1e6e49d90d30573a353a282c432fdee55